### DISCUSSION TITLE
constant CPU utilization

### DISCUSSION THREAD

**Discussion by heapxor** [blakeblackshear/frigate#3686] [General]
hello,
i can see in my hassOS that process > (probably connected with frigate) pid 5137 is constantly utilizing ~20% of cpu, even i have google coral.

any idea if thats bug or config issue?

thanks!

![image](https://user-images.githubusercontent.com/64676/185766534-a6bb290a-1ea4-48ce-abca-21881ad674ea.png)

**Comment by NickM-27** [comment]
Frigate has more types of load than just object detection (which is what coral does). Do you have hardware acceleration enabled?

**Reply by heapxor** [reply]
what hw acceleration do you mean? i am running hass inside VM container (using ESXi 7) assume hw acc is enabled otherwise vm wont work/esxi

thanks!

**Reply by NickM-27** [reply]
Talking about hardware acceleration of decoding the video streams. 

https://docs.frigate.video/configuration/hardware_acceleration

**Reply by heapxor** [reply]
@NickM-27 not sure if i do understand subject... i have supermicro board there is no external graphic card.
the only setting i can see inside Esxi Is 

![image](https://user-images.githubusercontent.com/64676/185769256-13acf1ca-77ea-4db4-8218-bb02384f43f3.png)

and thats what i see inside hass vm
![image](https://user-images.githubusercontent.com/64676/185769266-848298e1-247d-4664-b1af-fee1a4cd2d96.png)


thanks

**Reply by NickM-27** [reply]
Okay, then you don't have integrated graphics and can't do hardware acceleration. 

It takes considerable resources to decode the video stream which explains the constant usage

**Reply by heapxor** [reply]
@NickM-27  by integrated graphic you mean? its board > https://www.supermicro.com/en/products/motherboard/x11ssl-f

well is it good idea then to buy some graphic card and connect it via pcie slot ? will it help offload cpu?

thanks

**Reply by NickM-27** [reply]
Does your intel cpu not have integrated graphics?

And yes, using a GPU for hardware acceleration reduces cpu usages by a lot

**Reply by heapxor** [reply]
@NickM-27 what do u mean? cpu with integrated graphic? never heard about that... y mean main board with integrated graphic?
its https://www.intel.com/content/www/us/en/products/sku/97474/intel-xeon-processor-e31230-v6-8m-cache-3-50-ghz/specifications.html

is there some suggested gpu that can help? 
thanks

**Reply by NickM-27** [reply]
No, all consumer CPUs have integrated graphics besides the F skus

https://www.cgdirector.com/best-cpus-with-integrated-graphics/

Xeon don't though.

**Reply by heapxor** [reply]
whats  F skus ?

in that case better idea would be to buy some pcie graphic card to the system? no clue which one :( or if that makes sense?

**Reply by NickM-27** [reply]
> whats  F skus ?
> 
> 
> 
> in that case better idea would be to buy some pcie graphic card to the system? no clue which one :( or if that makes sense?

F skus are intel CPUs that don't include integrated graphics. 

I can't make purchasing recommendations (is it worth it) as that's for you to decide. An AMD or Nvidia GPU will decrease cpu load via hardware acceleration

**Reply by heapxor** [reply]
okay let me rephrase that question:

1) currently frigate or that ffmpeg process is constantly utilizing my cpu to 670Mhz (permanent utilization). Assume 670 Mhz of cpu consumers for a simplicity 1W of power.
In case I add pcie vga that can handle hardware acceleration - does it mean that 670Mhz from cpu will move to vga and vga gpu will be constantly utilized and consume 1W instead of the cpu? Or because of the hw acceleration vga will somehow manage hw acceleration and will consume only 0.1W or something lower as does cpu?

2) alternative option, as my current machine doesnt have amd/nvidia gpu; bc its server supermicro board.
Is it possible to run frigate with coral lets say on Odroid XU4 to utilize usb3 port for coral ; and ffmpeg hw acceleration bc of gpu and somehow connect frigate View UI/module from Homeassistant to that frigate backend running on odroid?


thank you!

**Reply by NickM-27** [reply]
> okay let me rephrase that question:
> 
> 1) currently frigate or that ffmpeg process is constantly utilizing my cpu to 670Mhz (permanent utilization). Assume 670 Mhz of cpu consumers for a simplicity 1W of power.
> In case I add pcie vga that can handle hardware acceleration - does it mean that 670Mhz from cpu will move to vga and vga gpu will be constantly utilized and consume 1W instead of the cpu? Or because of the hw acceleration vga will somehow manage hw acceleration and will consume only 0.1W or something lower as does cpu?

There are way too many variables to be able to say how much percent or power usage will go down. Hardware acceleration is recommended for use with frigate as it takes load off the CPU. Many many variables dictate how much.

**Reply by heapxor** [reply]
@NickM-27 
what variables do you mean?
is it some rule of thumb what gpu can help? some basic one?

what about 
2)
alternative option, as my current machine doesnt have amd/nvidia gpu; bc its server supermicro board.
Is it possible to run frigate with coral lets say on Odroid XU4 to utilize usb3 port for coral ; and ffmpeg hw acceleration bc of gpu and somehow connect frigate View UI/module from Homeassistant to that frigate backend running on odroid?

**Reply by NickM-27** [reply]
Variables like camera resolution and frame rate, your memory speed and cpu load required to move frames to / from GPU memory, etc. 

Check out nvidia cards that support decode here https://developer.nvidia.com/video-encode-and-decode-gpu-support-matrix-new

As far as your second question, no idea if that odroid is supported but it will most likely perform worse

**Reply by heapxor** [reply]
@NickM-27  i see, 
maybe that one? 
https://www.gpuzoo.com/GPU-NVIDIA/GeForce_GT_1030_DDR4.html
but spec says PCI-E 3.0 x 16 and my mainboard says
1 PCI-E 3.0 x8 (in x16), 
so no idea whats x8 (in x16?) if it will be compatible?

also amd/ati might be better? nvidia seems to have some issue with driver... ?
seems i will have to remove frigate from HASS and install it in separate vm with frigate backend docker... and then somehow ? connect frigate UI from homeassistant from that frigate backend?

sorry for lame questions
thanks

**Reply by NickM-27** [reply]
> so no idea whats x8 (in x16?) if it will be compatible?

It is compatible just will be lower bandwidth, but that is not a problem because video decode is not bandwidth intensive.

> also amd/ati might be better? nvidia seems to have some issue with driver... ?

I am not sure with the exsi setup how that may cause complications, but in general AMD drivers are easier (but nvidia is not a problem in general as of recently)

> seems i will have to remove frigate from HASS and install it in separate vm with frigate backend docker... and then somehow ? connect frigate UI from homeassistant from that frigate backend?

Not sure why you have this idea? It will continue / can continue to work exactly the same as before just need to update the frigate.yml accordingly

**Reply by heapxor** [reply]
> Not sure why you have this idea? It will continue / can continue to work exactly the same as before just need to update the frigate.yml accordingly

the reason behind is simple; i will have to passthrough gpu card via ESXi to VM; hass vm and hassOS especially are very limited... so i am not sure if i am able setup nvidia drivers properly within homeassistant.

**Reply by NickM-27** [reply]
I see, that's fair. You just need nvidia drivers on your host, HA will pick up the nvidia card. 

If you're worried about that, AMD GPU will likely be easier

**Reply by heapxor** [reply]
@NickM-27  ah i forgot ... i want to use that gpu for frigate and plex ... so in that case yes, i have to have separate VM where i run frigate and plex ; in that case can i somehow then link frigate backend with homeassistant frigate UI?

thanks

**Reply by NickM-27** [reply]
> in that case can i somehow then link frigate backend with homeassistant frigate UI?

https://github.com/blakeblackshear/frigate-hass-addons/blob/main/frigate_proxy/README.md

**Comment by heapxor** [comment]
thats strange .. they say some hw acceleration is doable with kaby lake ... ie the one i have
https://github.com/kern2011/Freenas-Quicksync

**Reply by NickM-27** [reply]
Kaby lake personal CPUs do, but xeon regardless of generation do not

**Comment by ehn** [comment]
@heapxor I was in your situation with a Supermicro server and a Xeon CPU, lacking hardware acceleration that could help with video de-/encoding. I ended up getting a [PNY NVIDIA RTX A2000](https://www.pny.com/nvidia-rtx-a2000). With a bit of help (#3619) from @NickM-27, it was fairly straightforward to get it working on Ubuntu with Frigate running in Docker (not inside Home Assistant).

**Comment by heapxor** [comment]
hello,
i am very confused here ... so which card is recommended?:( i was also thinking to utilize it for Plex server;
for a decoding purpose and to offload frigate is it good way to go with GeForce GTX 1050 / 1050 Ti?

also would be nice if card can occupies only one slot? as on this photo... but its not 100% requirement...
![image](https://user-images.githubusercontent.com/64676/204416288-33cf785b-2659-45ea-88ec-963bcd370344.png)

thanks!

**Reply by NickM-27** [reply]
the docs have a link to nvidia docs on which GPUs support decode / encode and how many streams are supported https://docs.frigate.video/configuration/hardware_acceleration#nvidia-gpu

**Reply by heapxor** [reply]
@NickM-27 do u mean this link ? 
https://developer.nvidia.com/video-encode-and-decode-gpu-support-matrix-new 
i am looking into it but there are only codecs like mpeg h265 etc.. maybe i am looking to wrong place?

**Reply by NickM-27** [reply]
right, so choose a GPU that is on that list and also that supports the codecs your cameras use

**Comment by heapxor** [comment]
@NickM-27 sorry i read the links but i am still confused,
my camera is producing h264 or h265 stream, so to offload cpu ; the required GPU needs to decode h264/265 to MPEG-1 | MPEG-2 and this has to be supported by gpu?

thanks

**Reply by NickM-27** [reply]
No, there's no encoding to mpeg. You just want it to be able to decode h264 and h265 streams.

**Reply by heapxor** [reply]
@NickM-27 ah okay, and these streams (h264,265) cant be decoded by cpu instructions ... like gpu is required. got it.

**Reply by NickM-27** [reply]
they can be decoded on the CPU but that is a lot more work for the CPU since the hardware is general purpose meanwhile the GPU is built specifically for that.

**Reply by heapxor** [reply]
@NickM-27 got it.

so assuming that one can be used? PNY Quadro P400 Professional Graphics Card 2GB GDDR5 PCI Express 3.0 x16 Single Slot 3x Mini DisplayPort 5K Support Ultra Quiet Active Fan  (https://www.amazon.de/-/en/Professional-Graphics-Express-DisplayPort-Support/dp/B084GYZZGW/)  
and also if drivers in ubuntu will work with it?
thanks

**Reply by NickM-27** [reply]
Yes it should be capable of decoding the h264 and h265 streams. I don't think drivers will be an issue but of course can't say for sure

**Reply by heapxor** [reply]
hello, one off - the topic... assuming rpi4 doesnt come with gpu that can offload it? thx

**Reply by NickM-27** [reply]
yes rpi does have a GPU although it is currently broken https://github.com/blakeblackshear/frigate/issues/3780

**Reply by heapxor** [reply]
@NickM-27 maybe have to wait on fix or use older kernel? in that case i was thinking to run rpi4 + coral + gpu offload and frigate on top .. and link it via frigate gui on HAAS

**Reply by NickM-27** [reply]
if you were worried about 20% CPU usage then the rpi won't be a good fit, it may have a GPU but it is rather anemic compared to even an intel iGPU
